File: src/main/java/com/server/ordering/domain/Restaurant.java
package com.server.ordering.domain;

import com.server.ordering.domain.member.Owner;
import lombok.*;
import org.locationtech.jts.geom.Point;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

import static javax.persistence.CascadeType.REMOVE;
import static javax.persistence.FetchType.*;
import static javax.persistence.GenerationType.*;
import static lombok.AccessLevel.*;

@Entity
@Getter
@NoArgsConstructor(access = PROTECTED)
@RequiredArgsConstructor
public class Restaurant {

    @Id @GeneratedValue(strategy = IDENTITY)
    @Column(name = "restaurant_id")
    private Long id;

    @NonNull
    private String restaurantName;

    @NonNull
    private String ownerName;

    @OneToOne(mappedBy = "restaurant", fetch = LAZY)
    private Owner owner;

    @OneToMany(mappedBy = "restaurant", orphanRemoval = true)
    private List<Food> foods = new ArrayList<>();

    @OneToMany(mappedBy = "restaurant")
    private List<Order> orders = new ArrayList<>();

    @OneToMany(mappedBy = "restaurant", cascade = REMOVE, orphanRemoval = true)
    private List<RepresentativeMenu> representativeMenus = new ArrayList<>();

    @OneToMany(mappedBy = "restaurant", cascade = REMOVE, orphanRemoval = true)
    private List<Review> reviews = new ArrayList<>();

    @OneToMany(mappedBy = "restaurant", cascade = REMOVE, orphanRemoval = true)
    private List<Bookmark> bookmarks = new ArrayList<>();

    @NonNull
    @Column(name = "restaurant_address")
    private String address;

    @NonNull
    private Point location;

    @NonNull
    private Integer tableCount;

    @Version
    private int waitingCount;
    private int orderCount;

    @NonNull
    @Enumerated(value = EnumType.STRING)
    private FoodCategory foodCategory;

    @NonNull
    @Enumerated(value = EnumType.STRING)
    private RestaurantType restaurantType;

    private String profileImageUrl;
    private String backgroundImageUrl;
    private String notice;

    @NonNull
    private Integer orderingWaitingTime;
    @NonNull
    private Integer admissionWaitingTime;

    public void addFood(Food food) {
        this.foods.add(food);
        food.registerRestaurant(this);
    }

    public void putRestaurant(String restaurantName, String ownerName, String address,
                              int tableCount, FoodCategory foodCategory, RestaurantType restaurantType) {
        this.restaurantName = restaurantName;
        this.ownerName = ownerName;
        this.address = address;
        this.tableCount = tableCount;
        this.foodCategory = foodCategory;
        this.restaurantType = restaurantType;
    }

    public void putProfileImageUrl(String profileImageUrl) {
        this.profileImageUrl = profileImageUrl;
    }

    public void putBackgroundImageUrl(String backgroundImageUrl) {
        this.backgroundImageUrl = backgroundImageUrl;
    }

    public void putOrderWaitingTime(Integer minutes) { this.orderingWaitingTime = minutes; }

    public void putAdmissionWaitingTime(Integer minutes) { this.admissionWaitingTime = minutes; }

    public void addRepresentativeMenu(RepresentativeMenu menu) {
        representativeMenus.add(menu);
        menu.registerRestaurant(this);
    }

    public void registerReview(Review review) {
        this.reviews.add(review);
    }

    public void increaseWaitingCount() {
        this.waitingCount++;
    }

    public void increaseOrderCount() {
        this.orderCount++;
    }

    public void putNotice(String notice) {
        this.notice = notice;
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/domain/Restaurant.kt
package com.byeoru.ordering_server.domain

import com.byeoru.ordering_server.domain.member.Owner
import org.locationtech.jts.geom.Point
import java.util.ArrayList
import javax.persistence.*

@Entity
class Restaurant(restaurantName: String,
                 ownerName: String,
                 address: String,
                 location: Point,
                 tableCount: Int,
                 foodCategory: FoodCategory,
                 restaurantType: RestaurantType,
                 orderingWaitingTime: Int,
                 admissionWaitingTime: Int) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "restaurant_id")
     var id: Long? = null
         protected set

     var restaurantName: String = restaurantName
         protected set

     var ownerName: String = ownerName
         protected set

    @OneToOne(mappedBy = "restaurant", fetch = FetchType.LAZY)
     var owner: Owner? = null
         protected set

    @OneToMany(mappedBy = "restaurant", orphanRemoval = true)
     var foods: MutableList<Food> = ArrayList<Food>()
         protected set

    @OneToMany(mappedBy = "restaurant")
     var orders: MutableList<Order> = ArrayList<Order>()
         protected set

    @OneToMany(mappedBy = "restaurant", cascade = [CascadeType.REMOVE], orphanRemoval = true)
     var representativeMenus: MutableList<RepresentativeMenu> = ArrayList<RepresentativeMenu>()
         protected set

    @OneToMany(mappedBy = "restaurant", cascade = [CascadeType.REMOVE], orphanRemoval = true)
     var reviews: MutableList<Review> = ArrayList<Review>()
         protected set

    @OneToMany(mappedBy = "restaurant", cascade = [CascadeType.REMOVE], orphanRemoval = true)
    val bookmarks: List<Bookmark> = ArrayList<Bookmark>()

    @Column(name = "restaurant_address")
     var address: String = address
         protected set

     var location: Point = location
         protected set

     var tableCount: Int? = tableCount
         protected set

    @Version
     var waitingCount: Int = 0
         protected set

     var orderCount = 0
         protected set

    @Enumerated(value = EnumType.STRING)
     var foodCategory: FoodCategory = foodCategory
         protected set

    @Enumerated(value = EnumType.STRING)
     var restaurantType: RestaurantType = restaurantType
         protected set

     var profileImageUrl: String? = null
         protected set

     var backgroundImageUrl: String? = null
         protected set

     var notice: String? = null
         protected set

     var orderingWaitingTime: Int = orderingWaitingTime
         protected set

     var admissionWaitingTime: Int = admissionWaitingTime
         protected set

    fun addFood(food: Food) {
        foods.add(food)
        food.registerRestaurant(this)
    }

    fun putRestaurant(
        restaurantName: String, ownerName: String, address: String, location: Point,
        tableCount: Int, foodCategory: FoodCategory, restaurantType: RestaurantType
    ) {
        this.restaurantName = restaurantName
        this.ownerName = ownerName
        this.address = address
        this.location = location
        this.tableCount = tableCount
        this.foodCategory = foodCategory
        this.restaurantType = restaurantType
    }

    fun putProfileImageUrl(profileImageUrl: String?) {
        this.profileImageUrl = profileImageUrl
    }

    fun putBackgroundImageUrl(backgroundImageUrl: String?) {
        this.backgroundImageUrl = backgroundImageUrl
    }

    fun putOrderWaitingTime(minutes: Int) {
        orderingWaitingTime = minutes
    }

    fun putAdmissionWaitingTime(minutes: Int) {
        admissionWaitingTime = minutes
    }

    fun addRepresentativeMenu(menu: RepresentativeMenu) {
        representativeMenus.add(menu)
        menu.registerRestaurant(this)
    }

    fun registerReview(review: Review) {
        reviews.add(review)
    }

    fun increaseWaitingCount() {
        waitingCount++
    }

    fun increaseOrderCount() {
        orderCount++
    }

    fun putNotice(notice: String?) {
        this.notice = notice
    }
}